fix(queue): correct queue-size documentation guidance

anoldguy opened the pull request:
The documentation states the queue "only works when queue-size is bigger than available images." From my reading of the code, this appears backwards and may cause the issues reported in discussion #136.

When queue-size approaches or exceeds available images, the random selection loop struggles to find unseen images after several attempts, falling back to recent duplicates. Navigation also appears to hit buffer boundaries, blocking next and previous commands.

Updated guidance to recommend queue-size smaller than total images, ideally half or less. Added runtime warning when queue-size >= image count, matching the pattern used elsewhere for configuration validation.

Extracted helpers (`get_queue_and_file_count`, `warn_if_queue_too_large`) to avoid repeating logic across initialization and update paths.

Still learning Rust idioms, so feedback welcome on both the analysis and approach.

anoldguy committed:
fix(queue): correct queue-size documentation guidance

The documentation states the queue "only works when queue-size is bigger than available images." From my reading of the code, this appears backwards and may cause the issues reported in discussion #136.

When queue-size approaches or exceeds available images, the random selection loop struggles to find unseen images after several attempts, falling back to
recent duplicates. Navigation also appears to hit buffer boundaries, blocking next and previous commands.

Updated guidance to recommend queue-size smaller than total images, ideally half or less. Added runtime warning when queue-size >= image count, matching
the pattern used elsewhere for configuration validation.

Extracted helpers (`get_queue_and_file_count`, `warn_if_queue_too_large`) to avoid repeating logic across initialization and update paths.

Still learning Rust idioms, so feedback welcome on both the analysis and approach.

Addresses GitHub discussion #136.

anoldguy committed:
chore: run cargo fmt on queue warning changes

Apologies for missing this in the initial commit. I should have run `cargo fmt` before submitting the PR, I'm still getting used to the Rust workflow.

danyspin97 commented:
Hi @anoldguy and thank you for this PR as well :)

> The documentation states the queue "only works when queue-size is bigger than available images." From my reading of the code, this appears backwards and may cause the issues reported in discussion https://github.com/danyspin97/wpaperd/discussions/136.

You are right, it's backwards 😅 

> Updated guidance to recommend queue-size smaller than total images, ideally half or less. Added runtime warning when queue-size >= image count, matching the pattern used elsewhere for configuration validation.

Taking into account the current status of the queue implementation, it makes sense. However, I know that many people wants to avoid duplicates before having seen all wallpapers available, and this works against their use case. Instead of suggesting the queue to be half of the size of the available wallpapers, I would rather fix the code that is making the queue buggy. I am thinking that there should be option to avoid duplicated wallpapers without setting a number. What do you think?

anoldguy committed:
fix(queue): keep navigation in sync when an image is shown again

A repeated image was silently dropped on push, leaving the cursor on
a different image than the screen. From there the queue stopped
recording history and navigation drifted. Record a repeat by moving
it to the most recent slot; replays of history are left alone.

anoldguy committed:
fix(queue): remove the queue-size runtime warning

The warning steered users away from queue-size >= image count, yet
that is the setup most people want. Fix the queue instead of warning
about it; the selection fix lands next.

anoldguy committed:
fix(queue): never repeat a wallpaper until all have been shown

Rejection sampling could repeat an image by bad luck, and always did
once the queue held the whole collection. Sample directly from the
images not yet shown this cycle; when none remain, start a new cycle
that avoids only the image on screen. A queue at least as big as the
collection now cycles exactly as the README always described.

anoldguy committed:
fix(queue): make resize correct in every direction

Only shrinking a full buffer worked. Growing a buffer that was not
full left tail stale, so walking previous could index out of bounds
and crash; shrinking one never dropped entries, so eviction stopped
for good. The buffer is stored oldest to newest, which makes resize
a matter of dropping from the front until it fits.

anoldguy committed:
perf(queue): build the available set in one pass

Checking every file against the cycle window was quadratic at the
end of a cycle, which collection-sized queues hit on every change.
Collect the cycle into a set once and filter against it.

anoldguy committed:
feat(queue): size the queue to the collection by default

An unset queue-size now means the number of available wallpapers, so
nothing repeats until everything has been shown and there is no
number to tune. An explicit queue-size keeps the old window
behaviour. Single files and empty folders fall back to 10, and every
size clamps to at least 1.

anoldguy committed:
feat(queue): follow the folder as it changes

hotwatch already repopulates the file list when a wallpaper folder
changes; automatically sized queues now resize in the same event.
New images join the current cycle, removals drop the oldest history
first, and explicit sizes are left alone.

anoldguy committed:
Merge upstream/main into fix-queue-documentation

Upstream reworked the image picker while this branch was open: queue
navigation gained wrap semantics, indices left the picker API, and
wpaperctl set arrived with an action queue. Both sides had fixed
random selection for small collections; upstream rotates through the
queue, this branch reshuffles each cycle. The merge keeps upstream's
navigation and API shape, and this branch's selection, push, resize,
and automatic queue sizing. Wrap remains for walking back through
history; selection no longer wraps, so every pass through the
collection is shuffled afresh.

anoldguy commented on daemon/src/image_picker.rs:
```diff
@@ -660,44 +728,40 @@
+        .filter(|index| !shown.contains(&files[*index]))
+        .collect();
+    if !available.is_empty() {
+        return files[available[fastrand::usize(..available.len())]].to_path_buf();
```
@danyspin97 I'm pretty sure this is a safe use of `fastrand::usize()`, we're just picking a random wallpaper.

danyspin97 replied:
Yea, this is pretty safe for us :)
